### ISSUE TITLE
Protocol and TypeAlias don't seem to support TypeVarTuple concatenation

### ISSUE THREAD

**Issue by timabroad** [microsoft/pyright#6317]
**Describe the bug**
Concatentation of TypeVarTuples as described in [PEP 646](https://peps.python.org/pep-0646/#type-concatenation) doesn't seem to work for either Protocol or TypeAlias types.

**Code or Screenshots**
For the following code, my expectation would be that `test1`, `test2`, and `test3` should all effectively have the same type.
```python
from typing import Callable, Protocol, TypeAlias, TypeVarTuple, Unpack, reveal_type

Ts = TypeVarTuple('Ts')
ArgTs = TypeVarTuple('ArgTs')

class C(Protocol[Unpack[ArgTs]]):
    def __call__(self, *args: Unpack[ArgTs]) -> None: ...

D: TypeAlias = Callable[[Unpack[ArgTs]], None]

def test1(cb: Callable[[int, Unpack[Ts]], None]) -> tuple[Unpack[Ts]]: ...
def test2(cb: C[int, Unpack[Ts]]) -> tuple[Unpack[Ts]]: ...
def test3(cb: D[int, Unpack[Ts]]) -> tuple[Unpack[Ts]]: ...  # pyright considers this an error

def cb(a: int, b: str) -> None: ...

reveal_type(test1(cb)) # Correctly resolved as tuple[str]
reveal_type(test2(cb)) # pyright considers this an error
reveal_type(test3(cb)) # Incorrectly resolved as tuple[Unknown]
```

I get the following output from pyright:
```
test.py
  test.py:14:15 - error: Expected type expression but received "D[int, *Ts@test5]" (reportGeneralTypeIssues)
  test.py:18:13 - information: Type of "test3(cb3)" is "tuple[str]"
  test.py:19:19 - error: Argument of type "(a: int, b: str) -> None" cannot be assigned to parameter "cb" of type "C[int, *Ts@test4]" in function "test4"
    Type "(a: int, b: str) -> None" cannot be assigned to type "(int, *Ts@test4) -> None" (reportGeneralTypeIssues)
  test.py:19:13 - information: Type of "test4(cb3)" is "tuple[Unknown]"
  test.py:20:13 - information: Type of "test5(cb3)" is "tuple[Unknown]"
2 errors, 0 warnings, 3 informations
```

Mypy also doesn't like the Protocol variant, but does get the TypeAlias variant correctly:
```
test.py:18: note: Revealed type is "tuple[builtins.str]"
test.py:19: note: Revealed type is "builtins.tuple[<nothing>, ...]"
test.py:19: error: Argument 1 to "test2" has incompatible type "Callable[[int, str], None]"; expected "C[int, Unpack[tuple[<nothing>, ...]]]"  [arg-type]
test.py:20: note: Revealed type is "tuple[builtins.str]"
```

**VS Code extension or command-line**
Using `pyright 1.1.334` at the command line as well as in VS using `Pylance language server 2023.7.10 (pyright 74d8f3c4)`.

**Comment by timabroad** [comment]
A slightly more complicated example using multiple TypeVarTuples:
```python
from typing import Callable, Generic, Protocol, TypeAlias, TypeVarTuple, Unpack, reveal_type

Ts = TypeVarTuple('Ts')
ArgTs = TypeVarTuple('ArgTs')

class C(Protocol[Unpack[ArgTs]]):
    def __call__(self, *args: Unpack[ArgTs]) -> None: ...

class Test(Generic[Unpack[Ts]]):
    def test1(self, cb: C[Unpack[Ts], Unpack[ArgTs]]) -> tuple[Unpack[Ts], Unpack[ArgTs]]: ...
    def test2(self, cb: Callable[[Unpack[Ts], Unpack[ArgTs]], None]) -> tuple[Unpack[Ts], Unpack[ArgTs]]: ...

def cb(a: int, b: str, c: bytes, d: list[float]) -> None: ...

reveal_type(Test[int, str]().test1(cb))
reveal_type(Test[int, str]().test2(cb))
```

I'm not quite sure from this whether pyright supports multiple TypeVarTuple unpacks or not. In all cases there should only be one "unknown" TypeVarTuple that needs matching against, so in theory it should be possible to do the match. On the defintion of `test1()`, pyright complains about "error: Type argument list can have at most one unpacked TypeVarTuple or Tuple" - **but only for the return type and _not_ for the cb parameter**. On the definition of `test2()`, it complains about both the parameter and the return type, but it then it goes on to correctly deduce the type on the last line. This is the output I get:

```
test.py
  test.py:11:76 - error: Type argument list can have at most one unpacked TypeVarTuple or Tuple
  test.py:12:47 - error: Type argument list can have at most one unpacked TypeVarTuple or Tuple
  test.py:12:91 - error: Type argument list can have at most one unpacked TypeVarTuple or Tuple
  test.py:16:36 - error: Argument of type "(a: int, b: str, c: bytes, d: list[float]) -> None" cannot be assigned to parameter "cb" of type "C[int, str, *ArgTs@test1]"
in function "test1"
    Type "(a: int, b: str, c: bytes, d: list[float]) -> None" cannot be assigned to type "(int, str, *ArgTs@test1) -> None"
      Function accepts too few positional parameters; expected 4 but received 3 (reportGeneralTypeIssues)
  test.py:16:13 - information: Type of "Test[int, str]().test1(cb)" is "tuple[int, str, Unknown]"
  test.py:17:13 - information: Type of "Test[int, str]().test2(cb)" is "tuple[int, str, bytes, list[float]]"
4 errors, 0 warnings, 2 informations
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.335, which I just published. It will also be included in a future release of pylance.

**Comment by timabroad** [comment]
Great stuff, thank you!
